new branch

bgMargarita opened the pull request:
delete strings
edit variable

bgMargarita committed:
new branch
delete strings
edit variable

woainikk commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -3,23 +3,21 @@
             while ((line = readFromFile.readLine()) != null) {
-            mass.add(line);
+                Array.add(line);
                 System.out.println("gdfg");
```
зачем это?

bgMargarita replied:
название переменной поменяла

woainikk commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -3,23 +3,21 @@
-
     String line;
-    List<String> mass = new ArrayList<String>();
+    List<String> Array = new ArrayList<String>();
```
почему название листа с большой буквы? названия переменных с маленькой буквы пишем. для чего этот лист, можно осмысленное название ему дать?

AntipovAndrey replied:
https://www.oracle.com/technetwork/java/codeconventions-135099.html

bgMargarita replied:
окей, дадим

woainikk commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -3,23 +3,21 @@
 /**
  * Created by Batman on 27.11.2018.
  */
 public class Parser {
```
пустую строку после этой вставь

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -3,23 +3,21 @@
+    List<String> Array = new ArrayList<String>();
 
-    public void reader() {
+    public void readFromFile() {
```
Твой метод ничего не возвращает, а вместо этого сохраняет в себя что он прочитал из файла.
Было бы правильнее не делать поля
```
 String filename = "file.txt";
 List<String> Array = new ArrayList<String>();
```
А чтобы твой метод принимал название файла, и сам создавал List, и его возвращал. 
и работа с твоим классом будет такая
```
Parser p = new Parser();
List<String> l = p.readFromFile("filename.txt");
```

AntipovAndrey replied:
И по поводу try - catch - finally
Это правильно, что ты закрываешь файл после использования, но с 7 джавы появилась возможность делать это проще
сейчас ты написала
```
BufferedReader readFromFile = null;
try {
   ...
} catch(IOException e) {
  ...
} finally {
   (закртыие файла)
}
```
А можно было 
```
try (BufferedReader readFromFile = new BufferedReader(new FileReader("file.txt"))) {
 ...
} catch (IOExceprion e) {
 ...
}
```
В итоге джава сама за тебя закроет файл при выходе из try-catch блока, и такой способ предпочтительнее

woainikk commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -3,23 +3,21 @@
-    public void reader() {
+    public void readFromFile() {
         BufferedReader readFromFile = null;
         try {
```
try with resources

bgMargarita committed:
remove String filename
made ReadFromFile return List
remove file.close()
start postfix translation(need to do priority choice)

bgMargarita committed:
add string after parser initialising

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,40 +1,65 @@
+        }
+    }
+
+    public List<String> ReadFromFile(String filename) {
```
именование опять не соболюдено 
с большой буквы только классы
функции, поля и переменные с маленькой

bgMargarita replied:
ай, извините

woainikk commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,41 +1,66 @@
+            else if (str.get(i) == "(")
+                stack.push(str.get(i));
+            else if (str.get(i) == ")") {
+                stack.push(str.get(i));
```
обединить два ифа со скобками в один

AntipovAndrey replied:
Строки на == не проверяют
надо str.get(i).equals(")")

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,41 +1,66 @@
+    }
+
+    public void WriteToStack(List<String> str) {
+        Stack stack = new Stack();
```
Если ты знаешь какой тип ты хочешь хранить в стеке, то надо это указывать - 
```
Stack<String>
```
Но и Stack это устаревшая коллеция, Сейчас исаользуется
```
Deque<String> stack = new ArrayDeque<>();
```

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,41 +1,66 @@
+    private static final Map<String, Integer> OPERATION_PRIORITY = new HashMap<String, Integer>();
+
+    static {
+        OPERATION_PRIORITY.put("*", 1);
```
Вся логика преобразования в постфиксную запись не очень окей держать в парсере. Задача парсера просто отдеать тебе строку с выражением (или список строк с выражениями), 
дальше ты эту строку отдаешь уже калькулятору, типа 
```
calculator.fromPostfix(parsedString);
```

bgMargarita committed:
remove capital letter mistake

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,41 +1,66 @@
         BufferedReader readFromFile = null;
         try {
-            readFromFile = new BufferedReader(new FileReader("\\Users\\Batman\\IdeaProjects\\Calculator\\src\\file.txt"));
+            readFromFile = new BufferedReader(new FileReader(filename));
```
Теперь ты вообще не закрываешь файл никак. 
надо написать 
```
try (BufferedReader readFromFile = new BufferedReader(new FileReader(filename))) {
}
```
А сверху это удаляешь
```
BufferedReader readFromFile = null;
```

bgMargarita committed:
divide functions into classes Calculator and Parser

woainikk commented on src/main/java/com/Margarita/calculator/Calculator.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Created by Batman on 28.11.2018.
+ */
+public class Calculator {
```
вставить пустую строку

woainikk commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,41 +1,34 @@
-                System.out.println("gdfg");
+    private List<String> stringToArray(String str) {
+        return Arrays.asList(str.split(" "));
+
```
удалить пустую строку

woainikk commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,41 +1,34 @@
+        return stringToArray(stringsFromFile.get(stringsFromFile.size() - 1));
     }
 
+
```
удалить пустую строку

bgMargarita committed:
finished translation to Postix

woainikk commented on src/main/java/com/Margarita/calculator/Calculator.java:
```diff
@@ -2,39 +2,52 @@
+    private int getPriority(String operator) {
+        if (operator.equals("+") || operator.equals("-"))
+            return 2;
+        else if (operator.equals("*") || operator.equals("/"))
```
зачем тут else ?

bgMargarita committed:
remove else

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Calculator.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public class Calculator {
+
+    private int calculate(Integer operand1, Integer operand2, String operator) {
```
Я думаю все же, что ты не только с интами можешь работать

AntipovAndrey replied:
Я думаю будет правильнее завести интерфейс Operator (или можно назвать Function), 
там можно определить один метод, например evaluate
Она принимает два double
```java
interface Operator {
    double call(double a, double b); 
}
```
Дальше какая схема - ты делаешь несколько имплементаций (классы оч короткими выйдут)
```java
class PlusOperator implements Operator {

    @Override
    public double evaluate(double a, double b) {
        return a + b;
    }
}
```
После создания основных операторов ты делаешь мап в калькуляторе
и делаешь что-то вроде
```java
operatirs.put("+", new PlusOperator());
operatirs.put("-", new MinusOperator());
```
И далее, твой метод будет выглядить очень коротко
```java
double calculate(double operand1, double operand2, String operator) {
    operators.get(operator).evaluate(operand1, operand2)
}
```
Огромный плюс такого подхода в том, что можно будет добавить новые операторы всего-лишь добавив в мапу новую реализацию того интерфейса, и все само заработает

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Calculator.java:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+        return stack.peek();
+
```
обычно про кодстайл не пишу, но не надо оставлять дыры в коде, 
отделять путой сторокой класс и его поля это ок, но в методе не надо такого
в методе окей отделять какие-то логические куски (но обычно такое вообще можно на более мелкие методы разбить)

AntipovAndrey commented on src/main/java/com/Margarita/calculator/Parser.java:
```diff
@@ -1,41 +1,31 @@
 
-    String line;
-    List<String> mass = new ArrayList<String>();
+    private String line;
```
не думаю, что тебе тут нужно именно поле, если оно используется только в одном методе класса. Можно сделать обычной переменной

bgMargarita committed:
small mistakes removed

bgMargarita committed:
small mistakes removed

bgMargarita committed:
priority mistake fixed
new class Writer

bgMargarita committed:
fixed priorities

bgMargarita committed:
class Main
